ledger-service/http-json/src/main/scala/com/digitalasset/http/WebsocketEndpoints.scala

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) 2019 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.digitalasset.http
+
+import akka.http.scaladsl.model.HttpMethods._
+import akka.http.scaladsl.model._
+import akka.http.scaladsl.model.ws.{Message, UpgradeToWebSocket}
+import akka.stream.Materializer
+import akka.stream.scaladsl.Flow
+import com.digitalasset.jwt.domain.Jwt
+import com.digitalasset.ledger.api.refinements.{ApiTypes => lar}
+import com.typesafe.scalalogging.StrictLogging
+import scalaz.syntax.std.option._
+import scalaz.{-\/, \/, \/-}
+import scala.concurrent.{ExecutionContext, Future}
+import EndpointsCompanion._
+
+object WebsocketEndpoints {
+  private val tokenPrefix: String = "jwt.token."
+  private val wsProtocol: String = "daml.ws.auth"
+}
+
+class WebsocketEndpoints(ledgerId: lar.LedgerId,
+                         decodeJwt: ValidateJwt,
+                         webSocketService: WebSocketService)
+                        (implicit mat: Materializer, ec: ExecutionContext) extends StrictLogging {
+
+  import WebsocketEndpoints._
+
+  lazy val transactionWebSocket: PartialFunction[HttpRequest, Future[HttpResponse]] = {
+    case req@HttpRequest(GET, Uri.Path("/transaction/connect"), _, _, _) =>
+      req.header[UpgradeToWebSocket] match {
+        case Some(upgradeReq) =>
+          connect(upgradeReq, Some(wsProtocol))
+        case None => Future.successful(httpResponseError(InvalidUserInput(s"Cannot upgrade client's connection to websocket")))
+      }
+  }
+
+  private[http] def connect(req: UpgradeToWebSocket, protocol: Option[String]): Future[HttpResponse] = {
+    protocol.map(req.requestedProtocols.contains(_)) match {
+      case Some(valid) if valid => handleWebsocketRequest(req, protocol)
+      case _ => Future.successful(httpResponseError(Unauthorized(s"Missing required $tokenPrefix.[token] or $wsProtocol subprotocol")))
+    }
+  }
+
+
+  private def handleWebsocketRequest(req: UpgradeToWebSocket, protocol: Option[String]):Future[HttpResponse] = {
+    findJwtFromSubProtocol(req).flatMap(decodeAndParsePayload(_, decodeJwt)) match {
+      case \/-((jwt, jwtPayload)) =>
+        val handler: Flow[Message, Message, _] = webSocketService.transactionMessageHandler(jwt, jwtPayload)
+        Future.successful(req.handleMessages(handler, protocol))
+      case -\/(e) => Future.successful(httpResponseError(e))
+    }
+  }
+
+  private def findJwtFromSubProtocol(upgradeToWebSocket: UpgradeToWebSocket): Unauthorized \/ Jwt = {
+    upgradeToWebSocket.requestedProtocols.collectFirst {
```

Comment:
Note: here we use subprotocols to find jwts, it should be in this format: `jwt.token.{bearer_token}, daml.ws.auth`
This `daml.ws.auth` is necessary to pass along the handshake to a specific subprotocol we accept. cc @bame-da @leo-da